src/ed25519/keyring.rs

```diff
@@ -72,4 +59,40 @@ impl KeyRing {
             Ok(())
         }
     }
+
+    // TODO(ismail): figure out where we get the pubkey from and delete this:
+    pub fn sign_with_only_signer(msg: &[u8]) -> Result<Ed25519Signature, Error> {
+        let keyring = GLOBAL_KEYRING.read().unwrap();
+
+        let mut vals = keyring.0.values();
+        if vals.len() > 1 {
+            return Err(err!(SigningError, "expected only one key in keyring"));
+        }
+        let signer = vals.next().unwrap();
+
+        signer.sign(msg)
+    }
+
+    pub fn get_only_signing_pubkey() -> Result<PublicKey, Error> {
+        let keyring = GLOBAL_KEYRING.read().unwrap();
+
+        let mut keys = keyring.0.keys();
+        if keys.len() > 1 {
+            return Err(err!(InvalidKey, "expected only one key in keyring"));
+        }
+        Ok(*keys.next().unwrap())
+    }
+
+    /// Sign a message using the secret key associated with the given public key
+    /// (if it is in our keyring)
+    pub fn sign(public_key: &PublicKey, msg: &[u8]) -> Result<Ed25519Signature, Error> {
```

Comment:
An alternative signature for this which can handle both cases without splitting them into separate functions:

```rust
pub fn sign(public_key: Option<&PublicKey>, msg: &[u8]) -> Result<Ed25519Signature, Error>
```